test/micro/org/openjdk/bench/java/lang/foreign/CriticalCalls.java

```diff
@@ -0,0 +1,113 @@
+
+import java.lang.foreign.Arena;
+import java.lang.foreign.FunctionDescriptor;
+import java.lang.foreign.Linker;
+import java.lang.foreign.MemorySegment;
+import java.lang.foreign.SegmentAllocator;
+import java.lang.foreign.SymbolLookup;
+
+import java.lang.invoke.MethodHandle;
+
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.util.concurrent.TimeUnit;
+
+import static java.lang.foreign.ValueLayout.*;
+
+@BenchmarkMode(Mode.AverageTime)
+@Warmup(iterations = 5, time = 500, timeUnit = TimeUnit.MILLISECONDS)
+@Measurement(iterations = 10, time = 500, timeUnit = TimeUnit.MILLISECONDS)
+@State(org.openjdk.jmh.annotations.Scope.Thread)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@Fork(value = 3, jvmArgsAppend = { "--enable-native-access=ALL-UNNAMED" })
+public class CriticalCalls {
+
+    static final MethodHandle PINNED;
+    static final MethodHandle NOT_PINNED;
+
+    static {
+        System.loadLibrary("CriticalCalls");
+        SymbolLookup lookup = SymbolLookup.loaderLookup();
+
+        MemorySegment sumIntsSym = lookup.find("sum_ints").get();
+        FunctionDescriptor sumIntsDesc = FunctionDescriptor.of(JAVA_INT, ADDRESS, JAVA_INT);
+
+        PINNED = Linker.nativeLinker().downcallHandle(
+            sumIntsSym,
+            sumIntsDesc,
+            Linker.Option.critical(true));
+        NOT_PINNED = Linker.nativeLinker().downcallHandle(
+            sumIntsSym,
+            sumIntsDesc,
+            Linker.Option.critical(false));
+    }
+
+    @Param({"100", "10000", "1000000"})
+    int size;
+
+    int[] arr;
+    SegmentAllocator recycler;
+
+    @Setup
+    public void setup() {
+        arr = new int[size];
+        for (int i = 0; i < size; i++) {
+            arr[i] = i;
+        }
+
+        recycler = SegmentAllocator.prefixAllocator(Arena.ofAuto().allocate(JAVA_INT, arr.length));
+    }
+
+    @Benchmark
+    public int callPinned() throws Throwable {
+        return (int) PINNED.invokeExact(MemorySegment.ofArray(arr), arr.length);
+    }
+
+    @Benchmark
+    public int callNotPinned() throws Throwable {
+        try (Arena arena = Arena.ofConfined()) {
+            MemorySegment nativeArr = arena.allocate(JAVA_INT, arr.length);
+            MemorySegment.copy(arr, 0, nativeArr, JAVA_INT, 0, arr.length);
+            return (int) NOT_PINNED.invokeExact(nativeArr, arr.length);
+        }
+    }
+
+    @Benchmark
+    public int callRecycled() throws Throwable {
+        MemorySegment nativeArr = recycler.allocate(JAVA_INT, arr.length);
+        MemorySegment.copy(arr, 0, nativeArr, JAVA_INT, 0, arr.length);
```

Comment:
Same here - I believe a single call to the correct `allocate` method suffices, but in this case I don't think performance should be skewed too much.